### ISSUE TITLE
Carriers with no fighters left must not be considered "armed"

### ISSUE THREAD

**Issue by Vezzra** [freeorion/freeorion#2343]
### Environment

* **FreeOrion Version**: bd966bf9d5278562d73a2236804666386f985401

### Description

A carrier that had its entire fighter wing destroyed in previous combats and no other weapons apparently is still considered "armed" by the engine.

The most problematic consequence of this is that a single stealthed carrier can enforce a blockade even if it has lost all its fighters, resulting in an essentially blockade force (until the enemy empires manage to research the detection techs level necessary to detect the stealth carrier).

Brought up on the forum [here](https://www.freeorion.org/forum/viewtopic.php?p=94302#p94302).

### Expected Result

A carrier which has all its fighters destroyed and no other weaponry should not be considered "armed", and not be able to enforce blockades.

**Comment by agrrr3** [comment]
Did some code analysis: the armed condition relies on Ship::HasFighters, which delegates to the ship's design's ::HasFighters 

I am uncertain if the Ship::HasFighters should return true if there are no fighters left in the ship. 

So either the armed condition should ask more explicitly or HasFighters should be changed.

No matter which way is chosen: one should check all calls to Ship::HasFighters if the intention is to know the capacity to launch fighters or if there actually fighters still left.

**Comment by agrrr3** [comment]
Actually i think making Armed like suggested  might be impossible without bigger changes - i did the simple implementation and tested using stealth bombers against some monsters. At the end of the combat i had no fighters left, so with the new implementation it was Not Armed. But actually it gets resupplied and hits hard.

I think the implementation solves the blockade problem.
But i think if a ship is able to do damage it should match the Armed condition (that includes empty carriers resupplied with fighters).

**Comment by Vezzra** [comment]
> But i think if a ship is able to do damage it should match the Armed condition (that includes empty carriers resupplied with fighters).

Hm, an empty carrier which got resupplied with fighters isn't empty anymore... I guess we need to address the case when a carrier lost all its fighter in a battle, but is within supply, so will get its fighters reinforced immediately.

Which means we need to decide if a carrier should be considered "armed" in that case or not. Probably yes I guess?

**Comment by agrrr3** [comment]
A different solution to the blockade problem could be that we make fighters not able to enforce a blockade at all.

**Comment by Vezzra** [comment]
Not really a fan of that idea. Putting a rather serious restriction on a game mechanic for the sake of working around an issue in the engine instead of properly fixing the issue isn't how such issues should be dealt with.

**Comment by agrrr3** [comment]
The fighters-not-able-to-blockade is not intended a workaround. There is a
design gap. And it is one of the options to answer that design question.

So which of the following should be able to block?
Which of the following should match the Armed condition?
* carriers with at east one any type of fighter which does at least 1 damage
* carriers of fighters with 0 damage
* empty carrier which will not be resupplied
* empty stationary carriers which will be resupplied
* empty carriers which do hop out hop in in a single turn to get resupplied

**Comment by Vezzra** [comment]
> The fighters-not-able-to-blockade is not intended a workaround. There is a design gap. And it is one of the options to answer that design question.

Erm, as an option for a final solution to address the underlying design issue I like it even less. Why on earth should fighters not be able to enforce a blockade? That solution might be simple, but would be detrimental to gameplay IMO...

> So which of the following should be able to block?
> Which of the following should match the Armed condition? 

I'd vote for the first option:

> * carriers with at east one any type of fighter which does at least 1 damage

For simplicity sake we can decide that empty carriers (regardless if they can resupply their fighter wings until the next turn) can't keep up a blockade (the fluff explanation for that should be easy enough).

Basically, any ship that can do damage to enemy ships (by whatever means) should be counted as armed/be able to enforce a blockade.
